File: kern/kernel.py
"""Host-side microkernel model used for S6 bring-up.

This provides a stable syscall surface and boot flow for later VM program ports.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from kern.audio import SIG_AUDIO, AudioDriver
from kern.gpio import GpioController
from kern.input import SIG_HID, InputDriver
from kern.serial import SIG_UART, SerialModule
from kern.video import SIG_VIDEO, VideoDriver
from plover_vm.memory.bus import MemoryBus

SIG_FDD = 0xA1
SIG_GPIO = 0xC3

DRIVER_BY_SIG = {
    SIG_FDD: "vfdd",
    SIG_HID: "hid",
    SIG_AUDIO: "audio",
    SIG_VIDEO: "video",
    SIG_GPIO: "gpio",
    SIG_UART: "serial",
}


@dataclass
class KernelState:
    heap_base: int = 0x6100
    heap_ptr: int = 0x6100
    output: list[str] = field(default_factory=list)
    slot_signatures: dict[int, int] = field(default_factory=dict)
    device_table: dict[int, str] = field(default_factory=dict)


class Kernel:
    def __init__(self, bus: MemoryBus) -> None:
        self.bus = bus
        self.state = KernelState()
        self.gpio = GpioController()
        self.serial = SerialModule()
        self.video = VideoDriver(bus)
        self.audio = AudioDriver(bus)
        self.input = InputDriver(bus)

    def kprint(self, s: str) -> None:
        self.state.output.append(s)
        self.serial.write((s + "\n").encode("ascii", errors="replace"))
        self.video.print(s)

    def kmalloc(self, n: int) -> int:
        if n <= 0:
            return self.state.heap_ptr
        p = self.state.heap_ptr
        self.state.heap_ptr = (self.state.heap_ptr + n) & 0xFFFF
        return p

    def devmgr_scan(self, slots: int = 16) -> dict[int, str]:
        self.state.device_table = {}
        for idx in range(slots):
            sig = self.state.slot_signatures.get(idx, 0xFF) & 0xFF
            if sig in (0x00, 0xFF):
                continue
            drv = DRIVER_BY_SIG.get(sig, "unknown")
            self.state.device_table[idx] = drv
            self.kprint(f"DEV slot{idx} sig={sig:02X} drv={drv}")
        return dict(self.state.device_table)

    def boot(self) -> None:
        if not self.state.slot_signatures:
            # default teaching/demo topology: vfdd, gpio, serial
            self.state.slot_signatures = {
                0: SIG_FDD,
                1: SIG_GPIO,
                2: SIG_UART,
                3: SIG_VIDEO,
                4: SIG_AUDIO,
                5: SIG_HID,
            }
        self.kprint("kernel_boot")
        self.devmgr_scan()
        self.kprint("kernel_help")

File: crates/plover_os/src/kernel.rs
use plover_mmu::MemoryBus;
use std::collections::BTreeMap;

pub const SIG_FDD: u8 = 0xA1;
pub const SIG_GPIO: u8 = 0xC3;
pub const SIG_UART: u8 = 0xD4;
pub const SIG_VIDEO: u8 = 0x56;
pub const SIG_AUDIO: u8 = 0x41;
pub const SIG_HID: u8 = 0x48;

#[derive(Clone, Debug, Default)]
pub struct GpioController {
    pub direction: u8,
    pub port_a: u8,
}

impl GpioController {
    pub fn read_port(&self) -> u8 {
        self.port_a & 0xFF
    }

    pub fn set_input_bits(&mut self, mask: u8, values: u8) {
        self.port_a = (self.port_a & !mask) | (values & mask);
    }

    pub fn get_bit(&self, bit: u8) -> bool {
        (self.port_a >> bit) & 1 != 0
    }

    pub fn set_bit(&mut self, bit: u8) {
        self.port_a |= 1 << bit;
    }

    pub fn clear_bit(&mut self, bit: u8) {
        self.port_a &= !(1 << bit);
    }
}

#[derive(Clone, Debug, Default)]
pub struct SerialModule {
    pub signature: u8,
    pub tx_fifo: Vec<u8>,
    pub rx_fifo: Vec<u8>,
}

impl SerialModule {
    pub fn new() -> Self {
        Self {
            signature: SIG_UART,
            ..Default::default()
        }
    }

    pub fn status(&self) -> u8 {
        let mut st = 0x02; // ST_TX_READY
        if !self.rx_fifo.is_empty() {
            st |= 0x01; // ST_RX_READY
        }
        st
    }

    pub fn write(&mut self, data: &[u8]) {
        self.tx_fifo.extend_from_slice(data);
    }
}

#[derive(Clone, Debug, Default)]
pub struct KernelState {
    pub slot_signatures: BTreeMap<u8, u8>,
    pub device_table: BTreeMap<u8, String>,
    pub output: Vec<String>,
}

pub struct Kernel {
    pub bus: MemoryBus,
    pub state: KernelState,
    pub gpio: GpioController,
    pub serial: SerialModule,
}

impl Kernel {
    pub fn new(bus: MemoryBus) -> Self {
        Self {
            bus,
            state: KernelState::default(),
            gpio: GpioController {
                direction: 0x0F,
                port_a: 0,
            },
            serial: SerialModule::new(),
        }
    }

    pub fn kprint(&mut self, s: &str) {
        self.state.output.push(s.to_string());
        self.serial.write(&(s.to_string() + "\n").into_bytes());
        let mut text = s.to_string();
        text.push('\n');
        let bytes = text.as_bytes();
        let len = bytes.len().min(255) as u8;
        use plover_copro::vdu::CMD_VDU_PRINT;
        self.bus.mailbox.issue_vdu(CMD_VDU_PRINT, len, 0, Some(bytes));
    }

    pub fn devmgr_scan(&mut self) {
        self.state.device_table.clear();
        let sigs: BTreeMap<u8, u8> = if self.state.slot_signatures.is_empty() {
            [
                (0, SIG_FDD),
                (1, SIG_GPIO),
                (2, SIG_UART),
                (3, SIG_VIDEO),
                (4, SIG_AUDIO),
                (5, SIG_HID),
            ]
            .into_iter()
            .collect()
        } else {
            self.state.slot_signatures.clone()
        };
        for (idx, sig) in sigs {
            if sig == 0x00 || sig == 0xFF {
                continue;
            }
            let drv = match sig {
                SIG_FDD => "vfdd",
                SIG_GPIO => "gpio",
                SIG_UART => "serial",
                SIG_VIDEO => "video",
                SIG_AUDIO => "audio",
                SIG_HID => "hid",
                _ => "unknown",
            };
            self.state.device_table.insert(idx, drv.to_string());
            self.kprint(&format!("DEV slot{idx} sig={sig:02X} drv={drv}"));
        }
    }

    pub fn boot(&mut self) {
        if self.state.slot_signatures.is_empty() {
            self.state.slot_signatures = [
                (0, SIG_FDD),
                (1, SIG_GPIO),
                (2, SIG_UART),
                (3, SIG_VIDEO),
                (4, SIG_AUDIO),
                (5, SIG_HID),
            ]
            .into_iter()
            .collect();
        }
        self.kprint("kernel_boot");
        self.devmgr_scan();
        self.kprint("kernel_help");
    }
}
